src/lightning/fabric/utilities/distributed.py

```diff
@@ -383,3 +384,32 @@ def _distributed_is_initialized() -> bool:
     # https://github.com/pytorch/pytorch/blob/v2.1.0/torch/distributed/__init__.py#L25
     # this might happen to MacOS builds from source (default) or any build from source that sets `USE_DISTRIBUTED=0`
     return torch.distributed.is_available() and torch.distributed.is_initialized()
+
+
+class _InfiniteBarrier:
+    """A barrier with an infinite timeout.
+
+    Creates a new process group with the GLOO backend with a very high timeout that makes the barrier effectively wait
+    forever. This is useful in cases where you want to execute a long-running operation on a subset of ranks that should
+    not be subject to the regular collective timeout.
+
+    """
+
+    def __init__(self) -> None:
+        self.group = None
+        self.barrier = lambda: None
+
+    def __call__(self) -> None:
+        self.barrier()
+
+    def __enter__(self) -> Self:
+        if _distributed_is_initialized():
+            # Create a barrier with an 'infinite' timeout (only reliably possible over the GLOO backend)
+            self.group = torch.distributed.new_group(backend="gloo", timeout=timedelta(days=10000))
```

Comment:
Do you know if there could be problems from having initialized nccl and gloo groups concurrently? On cuda devices the strategy will have initialized the default nccl group

Reply:
The PyTorch docs don't mention any limitations, and the backend arg  in [torch.distributed.new_group](https://pytorch.org/docs/stable/distributed.html#torch.distributed.new_group) says this:

>  By default uses the same backend as the global group.

which suggests that this is definitely valid. I made several runs testing this out on CPU, GPUs, multi-node and it worked well so far. GLOO is always available regardless of the accelerator, so I'm confident it's good.